#

# Group 7 - Text Standartization & Encoding

This group work is designed to:

*   get you together and help you build your alumni network,
*   turn you into active contributors rather than passive consumers,
*   let you explain concepts in your own language,
*   produce high-quality documentation that can be used 1:1 as an exam cheat sheet
*  and, most importantly: have fun while doing serious machine learning.

There are no grades, no competition, and no bashing. The goal is not perfection — the goal is shared understanding.

# 1. Historical view: word vectors


### The basic problem: words vs numbers

Machine learning models cannot work directly with text.  
They require **numerical inputs**, so words must first be converted into numbers.

One of the earliest approaches was **one-hot encoding**, where each word is represented by a vector containing mostly zeros and a single one.

Example vocabulary:

| Word | Vector |
|------|--------|
| cat  | [1, 0, 0] |
| dog  | [0, 1, 0] |
| mat  | [0, 0, 1] |

This representation is simple but has a major limitation:  
it does **not capture meaning**.

For example, the distance between **cat** and **dog** is exactly the same as the distance between **cat** and **banana**, even though their meanings are clearly different.

This limitation motivated researchers to search for representations that could capture **semantic relationships between words**.

---

### The Distributional Hypothesis

A foundational idea from linguistics is the **distributional hypothesis**.

In the 1950s, **Zellig Harris** and **J. R. Firth** proposed that:

> Words that appear in similar contexts tend to have similar meanings.

Firth summarized this idea with the now-famous phrase:

> "You shall know a word by the company it keeps."

Consider the word **bank**. Its meaning shifts entirely depending on the surrounding context:

- *River, fish and flood* → **bank** refers to the side of a river
- *Loan, interest and deposit* → **bank** refers to a financial institution

The word is identical in both cases, but the context reveals its meaning. This is precisely what the distributional hypothesis captures.

A contradicting idea focused on grammar and syntax as the core of language — essentially saying meaning comes from formal rules.

---

## Co-occurrence and vector space models

To operationalize this idea, researchers started counting **how often words appear near each other** in large text corpora.

This produced **word–context co-occurrence matrices**, where each row describes the contexts in which a word appears.

Example:

| Word | sits | barks | mat |
|------|------|-------|-----|
| dog  | 3    | 8     | 1   |
| cat  | 4    | 0     | 6   |

Words with similar rows are assumed to have similar meanings.

This approach introduced the idea of representing words as **vectors in a high-dimensional space**.

---

## From sparse matrices to dense representations

However, these co-occurrence matrices quickly became **very large and sparse**, making them inefficient to use directly.

To solve this problem, researchers introduced **dimensionality reduction methods** such as **Latent Semantic Analysis (LSA)**.

LSA compresses the large matrix into a smaller vector space while preserving important semantic relationships.

This step was important because it showed that meaning could be captured using **dense numerical vectors instead of large count tables**.

---

### Word embeddings

Modern **word embeddings** represent each word as a dense vector of real numbers learned from large text corpora.

Example:

- cat → [0.21, -0.14, 0.73]  
- dog → [0.19, -0.10, 0.71]

Now similar words have similar vectors in the embedding space.

One famous example illustrating this property is:

**king − man + woman ≈ queen**

This shows that word vectors can capture **semantic relationships between words**.

![description](https://assets.zilliz.com/Figure_Word_Embeddings_b021a5a759.png)

---

### Historical importance

Word vectors represent a bridge between several traditions:

- linguistic theory (distributional semantics)
- statistical text analysis
- modern neural networks

They laid the foundation for later techniques such as **Word2Vec**, **GloVe**, and the contextual embeddings used in modern **large language models (LLMs)**.

---

## Takeaways

- Machine learning requires text to be converted into numerical representations.
- One-hot encoding was an early solution but does not capture meaning.
- The **distributional hypothesis** introduced the idea that context reveals meaning.
- Co-occurrence matrices were the first practical implementation of this idea.
- Dimensionality reduction methods like **LSA** produced dense semantic vectors.
- Modern **word embeddings** build on these ideas and form the basis of modern NLP.

# 2. Word2Vec

Word2Vec is a method for learning numerical vector representations (embeddings) of words from text data.
Instead of representing words as simple indices or one-hot vectors, Word2Vec trains a small neural network to predict relationships between words in a sentence.
During training, the model learns word vectors such that words appearing in similar contexts obtain similar vector representations.
These vectors capture semantic relationships between words and can be used as features in machine learning models.

The idea is based on the distributional hypothesis: words that occur in similar contexts tend to have similar meanings.

In [ ]:
!pip install gensim

In [ ]:
from gensim.models import Word2Vec
import nltk
from nltk.tokenize import word_tokenize

nltk.download('punkt')
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [ ]:
sentences_raw = [
    "the cat sits on the mat",
    "the dog sits on the floor",
    "the dog chased the cat",
    "cats and dogs are animals",
    "the cat climbed the tree"
]

sentences = [word_tokenize(s.lower()) for s in sentences_raw]

sentences

[['the', 'cat', 'sits', 'on', 'the', 'mat'],
 ['the', 'dog', 'sits', 'on', 'the', 'floor'],
 ['the', 'dog', 'chased', 'the', 'cat'],
 ['cats', 'and', 'dogs', 'are', 'animals'],
 ['the', 'cat', 'climbed', 'the', 'tree']]

## 2.1 Skip-gram

Skip-gram is a Word2Vec training method that predicts surrounding context words from a given center word.
For each word in a sentence, the model tries to predict the words that appear within a defined window around it.
By repeatedly performing this prediction task, the neural network learns vector representations (embeddings) in which words that occur in similar contexts have similar vectors.
Skip-gram typically performs well for rare words because each word generates multiple training examples.

Example sentence:

the cat sits on the mat

Training pairs with window size = 2:

Input: sits → Output: cat  
Input: sits → Output: on

In [ ]:
#Modell training based on the Sentences defined in Word2Vec above
model = Word2Vec(
    sentences,
    vector_size=50,   # dimension of word vectors
    window=2,         # context window size
    min_count=1,      # include rare words
    sg=1              # 1 = Skip-gram, 0 = CBOW
)

In [ ]:
# Vector of a word
model.wv["cat"]

array([-0.01631583,  0.0089916 , -0.00827415,  0.00164907,  0.01699724,
       -0.00892435,  0.009035  , -0.01357392, -0.00709698,  0.01879702,
       -0.00315531,  0.00064274, -0.00828126, -0.01536538, -0.00301602,
        0.00493959, -0.00177605,  0.01106732, -0.00548595,  0.00452013,
        0.01091159,  0.01669191, -0.00290748, -0.01841629,  0.0087411 ,
        0.00114357,  0.01488382, -0.00162657, -0.00527683, -0.01750602,
       -0.00171311,  0.00565313,  0.01080286,  0.01410531, -0.01140624,
        0.00371764,  0.01217773, -0.0095961 , -0.00621452,  0.01359526,
        0.00326295,  0.00037983,  0.00694727,  0.00043555,  0.01923765,
        0.01012121, -0.01783478, -0.01408312,  0.00180291,  0.01278507],
      dtype=float32)

In [ ]:
# How similar is a word to others
model.wv.most_similar("cat")
#cat
#dog
#climbed

[('sits', 0.12485622614622116),
 ('chased', 0.08061233162879944),
 ('climbed', 0.07399576157331467),
 ('the', 0.04237303137779236),
 ('tree', 0.018277151510119438),
 ('mat', 0.011398463509976864),
 ('dog', 0.011071980930864811),
 ('dogs', 0.0013422416523098946),
 ('floor', -0.012029263190925121),
 ('cats', -0.1094222441315651)]

In [ ]:
# The Vocubular of the Model based on its frequency
model.wv.index_to_key

['the',
 'cat',
 'dog',
 'on',
 'sits',
 'tree',
 'climbed',
 'animals',
 'are',
 'dogs',
 'and',
 'cats',
 'chased',
 'floor',
 'mat']

## 2.2 CBOW (Continuous Bag of Words)

CBOW is a **Word2Vec training method** that learns word embeddings by **predicting a target word from its surrounding context words**.

Example sentence:

```
The quick brown fox
```

CBOW training sample:

```python
["The", "quick", "fox"] -> "brown"
```

Interpretation:

```
"The quick ___ fox"
```

The model learns that the missing word is likely **brown**.

---

# Code Example 1 — Generate CBOW training samples

```python
text = "the quick brown fox jumps over the lazy dog"
tokens = text.split()

window = 2
samples = []

for i in range(window, len(tokens)-window):
    context = tokens[i-window:i] + tokens[i+1:i+window+1]
    target = tokens[i]
    samples.append((context, target))

samples
```

Example output:

```
(['the','quick','fox','jumps'], 'brown')
```

---

# Code Example 2 — Train Word2Vec with CBOW

```python
!pip install gensim
```

```python
from gensim.models import Word2Vec

sentences = [
    ["the","quick","brown","fox"],
    ["the","lazy","dog"],
    ["fox","is","quick"]
]

model = Word2Vec(
    sentences,
    vector_size=50,
    window=2,
    min_count=1,
    sg=0   # 0 = CBOW, 1 = Skip-gram
)
```

---

# Code Example 3 — Inspect embeddings

Word vector:

```python
model.wv["fox"]
```

Find similar words:

```python
model.wv.most_similar("fox")
```

---

# One-line summary

```
CBOW: context words → predict center word
Skip-gram: center word → predict context words
```

# 3. Byte-pair Encoding

**Byte-Pair Encoding (BPE)** was originally introduced in 1994 by Philip Gage as a straightforward and highly effective data compression technique. Long before its adaptation for modern AI and text tokenization, it was designed simply to reduce the file size of digital text and binary data by finding and replacing repeating patterns.


### Core Mechanism

The original algorithm operates through a greedy, iterative process of substitution. Here is how it works:

1. **Frequency Analysis:** The algorithm scans the entire dataset to identify the most frequently occurring pair of adjacent bytes.
2. **Substitution:** It replaces every instance of that frequent pair with a single, **unused byte** that does not currently appear anywhere in the data.
3. **Table Generation:** It records this substitution in a lookup table or dictionary (for example, documenting that the newly introduced byte `Z` represents the original pair `A` and `B`).
4. **Iteration:** The process repeats—scanning the newly compressed data for the next most frequent pair (which can include the newly substituted bytes)—until either no pairs appear more than once or all 256 possible byte values are in use.

To restore the original data, a decompression algorithm simply reads the compressed file and uses the generated lookup table in reverse, recursively expanding the substituted bytes back into their original pairs.

## BPE Step-by-Step Example

## Input String

```
aaabdaaabacbdaabafefgef
```

---

## Step 0: Initial State

```
String: a a a b d a a a b a c b d a a b a f e f g e f
```

Initial vocabulary: `{a, b, c, d, e, f, g}`

---

## Step 1: Find most frequent pair

Count all bigrams:

| Pair | Occurrences |
|------|-------------|
| **aa** | **4** |
| ab | 3 |
| bd | 2 |
| da | 2 |
| ba | 2 |
| ac | 1 |
| cb | 1 |
| af | 1 |
| fe | 2 |
| ef | 2 |
| fg | 1 |
| ge | 1 |

**Most frequent pair: `aa` (4 times)**

### → Merge `aa` → `Z`

```
Before: a a a b d a a a b a c b d a a b a f e f g e f
After:  Z a b d Z a b a c b d Z b a f e f g e f
```

> **Note:** `aaa` → `Z a` (greedy left-to-right: first two `a`'s merge, the third remains)

Vocabulary: `{a, b, c, d, e, f, g, Z(=aa)}`

---

## Step 2: Find most frequent pair

Current string: `Z a b d Z a b a c b d Z b a f e f g e f`

| Pair | Occurrences |
|------|-------------|
| **Za** | **2** |
| **ab** | **2** |
| bd | 2 |
| dZ | 1 |
| ba | 2 |
| ac | 1 |
| cb | 1 |
| Zb | 1 |
| af | 1 |
| fe | 2 |
| ef | 2 |
| fg | 1 |
| ge | 1 |

**Multiple pairs tied at 2.** We pick `Za` (first encountered).

### → Merge `Za` → `Y`

```
Before: Z a b d Z a b a c b d Z b a f e f g e f
After:  Y b d Y b a c b d Z b a f e f g e f
```

Vocabulary: `{a, b, c, d, e, f, g, Z(=aa), Y(=Za=aaa)}`

---

## Step 3: Find most frequent pair

Current string: `Y b d Y b a c b d Z b a f e f g e f`

| Pair | Occurrences |
|------|-------------|
| **Yb** | **2** |
| **bd** | **2** |
| dY | 1 |
| ba | 2 |
| ac | 1 |
| cb | 1 |
| dZ | 1 |
| Zb | 1 |
| af | 1 |
| fe | 2 |
| ef | 2 |
| fg | 1 |
| ge | 1 |

**Multiple pairs tied at 2.** We pick `Yb`.

### → Merge `Yb` → `X`

```
Before: Y b d Y b a c b d Z b a f e f g e f
After:  X d X a c b d Z b a f e f g e f
```

Vocabulary: `{a, b, c, d, e, f, g, Z(=aa), Y(=aaa), X(=Yb=aaab)}`

---

## Step 4: Find most frequent pair

Current string: `X d X a c b d Z b a f e f g e f`

| Pair | Occurrences |
|------|-------------|
| Xd | 1 |
| dX | 1 |
| Xa | 1 |
| ac | 1 |
| cb | 1 |
| bd | 1 |
| dZ | 1 |
| Zb | 1 |
| ba | 1 |
| af | 1 |
| **fe** | **2** |
| **ef** | **2** |
| fg | 1 |
| ge | 1 |

**Tied at 2: `fe` and `ef`.** We pick `fe`.

### → Merge `fe` → `W`

```
Before: X d X a c b d Z b a f e f g e f
After:  X d X a c b d Z b a W f g e f
```

> `f e f` → `W f` (left-to-right: first `fe` merges, the trailing `f` remains)

Vocabulary: `{a, b, c, d, e, f, g, Z(=aa), Y(=aaa), X(=aaab), W(=fe)}`

---

## Step 5: Find most frequent pair

Current string: `X d X a c b d Z b a W f g e f`

| Pair | Occurrences |
|------|-------------|
| Xd | 1 |
| dX | 1 |
| Xa | 1 |
| ac | 1 |
| cb | 1 |
| bd | 1 |
| dZ | 1 |
| Zb | 1 |
| ba | 1 |
| aW | 1 |
| Wf | 1 |
| fg | 1 |
| ge | 1 |
| ef | 1 |

**No pair occurs more than once → BPE stops.**

---

## Final Result Summary

| Step | Merge Rule | New Symbol | String After Merge |
|------|-----------|------------|-------------------|
| 0 | — | — | `aaabdaaabacbdaabafefgef` |
| 1 | `aa → Z` | Z | `ZabdZabacbdZbafefgef` |
| 2 | `Za → Y` | Y | `YbdYbacbdZbafefgef` |
| 3 | `Yb → X` | X | `XdXacbdZbafefgef` |
| 4 | `fe → W` | W | `XdXacbdZbaWfgef` |
| 5 | — (stop) | — | No pair with frequency > 1 |

### Final Compressed String

```
X d X a c b d Z b a W f g e f
```

### Final Vocabulary (Merge Table)

```
Z  = aa
Y  = Za  = aaa
X  = Yb  = aaab
W  = fe
```

### Decompression (Reverse)

To reconstruct the original string, apply merges in **reverse order**:

```
XdXacbdZbaWfgef
→ (W→fe)  XdXacbdZbafefgef
→ (X→Yb)  YbdYbacbdZbafefgef
→ (Y→Za)  ZabdZabacbdZbafefgef
→ (Z→aa)  aaabdaaabacbdaabafefgef  ✓
```

---

### Key Observations

1. **Greedy left-to-right:** When `aaa` is encountered, the first two `a`s merge into `Z`, leaving one `a` behind. The algorithm does **not** look ahead.

2. **Hierarchical merges build up:** Each new symbol can be part of the next merge — `aa→Z`, then `Za→Y`, then `Yb→X`. This builds a **hierarchy of subword units**: `X` = `aaab` was discovered bottom-up.

3. **Frequency-driven:** The algorithm always targets the most common pair, achieving maximum compression per step.

4. **In NLP context:** Instead of single characters, the initial "characters" are bytes or unicode characters, and instead of arbitrary letters like Z/Y/X, the merged tokens get integer IDs. The merge table is saved and reused at inference time to tokenize new, unseen text.

# 4. Modern encoding using LLMs

Large Language Models (LLMs) process language by converting text into numerical representations.
Since computers cannot directly understand human language, text must first be encoded into numbers.
This encoding process enables AI systems to analyze and interpret language.
Modern encoding techniques capture not only words but also their meaning and context.

### The Problem with Word2Vec

Modern encoding techniques rely on the Transformer architecture.
Transformers use a mechanism called self-attention.
Self-attention allows each word in a sentence to consider the other words around it.
This helps the model understand relationships and dependencies between words.

Word2Vec gives every word a **single fixed vector** — "bank" always maps to the same point in space, whether you mean a riverbank or a financial institution. The word's meaning is frozen at training time.

Modern LLMs solve this with **contextual embeddings**: the vector for a word changes depending on the sentence it appears in.

---

### The Core Idea

A transformer-based LLM reads the entire input sequence at once and produces one embedding vector **per token**, where each vector encodes not just the word but its role and meaning in that specific context.
```
"I deposited money at the bank."   → bank vector ≈ [finance region of space]
"We picnicked by the river bank."  → bank vector ≈ [nature region of space]
```

Same word, completely different vectors. This is the fundamental leap over Word2Vec.

### How It Works
```
Input text
    ↓
Tokenize  (split into subword tokens via BPE — see part 3. BPE)
    ↓
Token embeddings  (lookup table, like Word2Vec — still static at this stage)
    ↓
Transformer layers  (self-attention lets every token look at every other token)
    ↓
Contextual output embeddings  (one vector per token, now context-aware)
```

The transformer layers are where the magic happens. Self-attention allows "bank" to look at "river" or "money" in the same sentence and shift its representation accordingly.

In [ ]:
# 1. Load Model and Tokenizer
from transformers import AutoTokenizer, AutoModel
import torch

model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)
model.eval()

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
  

In [ ]:
# Tokenize the inputs

sentences = [
    "I deposited money at the bank.",
    "We picnicked by the river bank."
]

for sentence in sentences:
    tokens = tokenizer.tokenize(sentence)
    print(f"Sentence : {sentence}")
    print(f"Tokens   : {tokens}")
    print(f"Count    : {len(tokens)} tokens\n")

Sentence : I deposited money at the bank.
Tokens   : ['i', 'deposited', 'money', 'at', 'the', 'bank', '.']
Count    : 7 tokens

Sentence : We picnicked by the river bank.
Tokens   : ['we', 'picnic', '##ked', 'by', 'the', 'river', 'bank', '.']
Count    : 8 tokens



In [ ]:
# Extract Token-Level Embeddings

for sentence in sentences:
    inputs = tokenizer(sentence, return_tensors="pt")
    with torch.no_grad():
        outputs = model(**inputs)

    # last_hidden_state shape: [batch_size, num_tokens, hidden_size]
    token_embeddings = outputs.last_hidden_state[0]

    print(f"Sentence : {sentence}")
    print(f"Shape    : {token_embeddings.shape}  → (num_tokens, 768)")
    print(f"          = {token_embeddings.shape[0]} tokens, each a 768-dim vector\n")

Sentence : I deposited money at the bank.
Shape    : torch.Size([9, 768])  → (num_tokens, 768)
          = 9 tokens, each a 768-dim vector

Sentence : We picnicked by the river bank.
Shape    : torch.Size([10, 768])  → (num_tokens, 768)
          = 10 tokens, each a 768-dim vector



In [ ]:
# Extract the "bank" Vector from Each Sentence

def get_token_embedding(sentence, target_word, tokenizer, model):
    inputs = tokenizer(sentence, return_tensors="pt")
    tokens = tokenizer.convert_ids_to_tokens(inputs["input_ids"][0])

    with torch.no_grad():
        outputs = model(**inputs)

    token_embeddings = outputs.last_hidden_state[0]

    # Find the position of our target word
    token_idx = tokens.index(target_word)
    print(f"Tokens        : {tokens}")
    print(f"'{target_word}' found at position {token_idx}")
    return token_embeddings[token_idx]

bank_finance = get_token_embedding(sentences[0], "bank", tokenizer, model)
print()
bank_nature  = get_token_embedding(sentences[1], "bank", tokenizer, model)

Tokens        : ['[CLS]', 'i', 'deposited', 'money', 'at', 'the', 'bank', '.', '[SEP]']
'bank' found at position 6

Tokens        : ['[CLS]', 'we', 'picnic', '##ked', 'by', 'the', 'river', 'bank', '.', '[SEP]']
'bank' found at position 7


In [ ]:
# Compare the Two "bank" Vectors

from torch.nn.functional import cosine_similarity

sim = cosine_similarity(bank_finance.unsqueeze(0), bank_nature.unsqueeze(0)).item()
print(f"Cosine similarity between 'bank' (finance) and 'bank' (nature): {sim:.4f}")

# For reference: what do similarity scores mean?
print("\nReference scale:")
print("  1.00 = identical vectors (same word, same context)")
print("  0.90+ = very similar meaning")
print("  0.70  = related but distinct")
print("  <0.50 = clearly different meanings")

Cosine similarity between 'bank' (finance) and 'bank' (nature): 0.5455

Reference scale:
  1.00 = identical vectors (same word, same context)
  0.90+ = very similar meaning
  0.70  = related but distinct
  <0.50 = clearly different meanings


Interpretation: the two "bank" vectors are moderately similar — they share surface-level word identity — but BERT has pushed them apart in embedding space because their surrounding context words ("money" vs "river") signal different meanings.

Compare this to Word2Vec: it would give cosine similarity = 1.00 for the same word, because Word2Vec has only ONE vector per word regardless of context.

In [ ]:
# Sentence-Level Embedding via Mean Pooling

def get_sentence_embedding(sentence, tokenizer, model):
    inputs = tokenizer(sentence, return_tensors="pt")
    with torch.no_grad():
        outputs = model(**inputs)
    # Average all token vectors → one vector for the whole sentence
    return outputs.last_hidden_state.mean(dim=1).squeeze()

emb1 = get_sentence_embedding(sentences[0], tokenizer, model)
emb2 = get_sentence_embedding(sentences[1], tokenizer, model)

sim = cosine_similarity(emb1.unsqueeze(0), emb2.unsqueeze(0)).item()
print(f"Sentence embedding shape : {emb1.shape}  → one 768-dim vector per sentence")
print(f"Sentence-level similarity: {sim:.4f}")

Sentence embedding shape : torch.Size([768])  → one 768-dim vector per sentence
Sentence-level similarity: 0.6604


The sentences are more similar at the sentence level than at the word level — they share structure, topic (bank), and grammar. Mean pooling blends all token signals together, smoothing out the word-level difference we saw in Step 5.

This is a common trade-off: sentence embeddings lose token-level precision but are practical for tasks like semantic search or classification.

### Word2Vec vs. Modern LLM Encodings

| Property | Word2Vec | LLM (e.g. BERT) |
|---|---|---|
| Embedding per word | One fixed vector | Different vector per context |
| Handles polysemy | ✗ | ✓ |
| Captures syntax | Weakly | Strongly (via attention) |
| Requires fine-tuning for tasks | ✓ (usually) | Often works out of the box |
| Computational cost | Low | High |

### Key Takeaways

- **Word2Vec**: one word = one vector, always. "bank" is always the same point in space no matter what sentence it's in.
- **BERT**: one word = different vector depending on context. "bank" near "money" lands in a different region of space than "bank" near "river". Same spelling, different meaning, different vector.
- Every token gets its own 768-number description. That's the embedding. More numbers = more room to encode nuance.
- `[CLS]` and `[SEP]` are special tokens BERT adds automatically — they're not your words, but they take up slots in the output.
- To go from token embeddings → one vector for the whole sentence: just average all token vectors. This is called **mean pooling**.
- **Cosine similarity** tells you how close two vectors are in meaning: 1.0 = identical, 0.0 = unrelated. The two "bank" sentences score ~0.55 — contextually different, which is exactly what we want.
- BERT reads the whole sentence at once (left AND right context). GPT reads left to right only. That's why BERT is preferred for understanding/embeddings and GPT for generating text.

---

## Overall Takeaways

This notebook traces a single idea from its origins to modern practice: **how do we give a machine a useful numerical representation of a word?**

Each section answers a limitation of the previous one:

**One-hot encoding** (Section 1) was the naive starting point — simple, but completely blind to meaning. "Cat" and "dog" are just as far apart as "cat" and "volcano."

**Word2Vec** (Section 2) was the first real breakthrough. Instead of hand-crafting vectors, the model learns them from data by exploiting context. Skip-gram and CBOW are two strategies for the same goal: words that appear in similar contexts should end up in similar regions of vector space. The result — vectors where `king − man + woman ≈ queen` — showed that meaning has geometry.

**BPE** (Section 3) solved a practical problem that Word2Vec ignored: what do you do with words you have never seen before? By breaking words into reusable subword pieces, BPE ensures the vocabulary stays manageable and unknown words can still be represented. This is now the standard tokenisation strategy in every modern LLM.

**Contextual embeddings with BERT** (Section 4) closed the final gap. Word2Vec gives every word one fixed vector for life — "bank" is always the same point regardless of whether it sits next to "river" or "interest rate." BERT generates a fresh vector for every word in every sentence, shaped by what surrounds it. Same word, different meaning, different vector.

**The through-line:** we went from counting to learning, from whole words to subwords, and from static to contextual. Each step added expressiveness at the cost of complexity. Modern NLP lives at the BERT end of this spectrum — but understanding all four steps is what makes the final result legible rather than magical.

Authors:    
   1. Artemii Ponomarenko
   2. Daniela-Valentina Otelea
   3. Kaj Rauch
   4. Linus-Bertil Schmid
   5. Robin Schneider
   6. Sachin Varma
   7. Janis Kramer
   8. Céline Spirig
   9. Florian Givel
  10. Vojtech Marecek